### ISSUE TITLE
Lint request: warn when temporaries are used for values for which dropping has side effects

### ISSUE THREAD

**Issue by PieterPenninckx** [rust-lang-nursery/rust-clippy#1904]
Temporary values (that I tend to call 'anonymous values') are values for which there is no variable binding. E.g.: 

```rust
let a = Vec::<u8>::new().len();
```
Here the value of the type `Vec<u8>` is a temporary value.

Temporary values are -- generally speaking -- dropped at the semi-colon that terminates the statement in which they were created. More details can be found in [the reference](https://doc.rust-lang.org/stable/reference/expressions.html#temporary-lifetimes). This is not so well known. This can lead to bugs when a temporary is used to store a value for which the precise point where it is dropped is important, such as a `LockResult`. An example can be found in rust-lang/rust#37612. 

To prevent such bugs, I think it would be a good idea if clippy would warn when a 'sensitive' value is stored in a temporary. With a 'sensitive' value, I mean something like a `LockResult` where calling the `drop()` has important side effects (so not just releasing memory).

**Comment by PieterPenninckx** [comment]
I realized that my request was too vague about which datatypes have important side effects when being dropped. Since there is no way clippy could judge this, I would just work with a list. This will not capture all problems, but hopefully enough of them. This is the list I came up with:

* std::sync::LockResult&lt;T&gt;
* std::sync::TryLockResult&lt;T&gt;
* std::sync::MutexGuard&lt;T&gt;
* std::sync::RwLockReadGuard&lt;T&gt;
* std::sync::RwLockWriteGuard&lt;T&gt;
* std::cell::Ref&lt;T&gt;
* std::cell::RefMut&lt;T&gt;

**Comment by PieterPenninckx** [comment]
I'm working on this.

**Comment by academician** [comment]
Is it worth revisiting and possibly closing this issue after the [changes to temporary scope](https://doc.rust-lang.org/edition-guide/rust-2024/temporary-tail-expr-scope.html) in the 2024 edition?
